cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java

```diff
@@ -99,35 +103,48 @@ public class DefaultAccessEnforcer extends AbstractAccessEnforcer {
   @Override
   public void enforce(EntityId entity, Principal principal, Set<? extends Permission> permissions)
       throws AccessException {
+    EnforcementResult enforcementResult = enforceWithResult(entity, principal, permissions);
+    if (enforcementResult.isAuthorized().equals(EnforcementResult.AuthorizedResult.NO)) {
+      throw new UnauthorizedException(principal, permissions, entity);
+    }
+  }
+
+  public EnforcementResult enforceWithResult(EntityId entity, Principal principal,
+                                             Set<? extends Permission> permissions)
+    throws AccessException {
+
     MetricsContext metricsContext = createEntityIdMetricsContext(entity);
+    EnforcementResult enforcementResult;
     if (internalAuthEnabled && principal.getFullCredential() != null
         && principal.getFullCredential().getType() == Credential.CredentialType.INTERNAL) {
       LOG.trace("Internal Principal enforce({}, {}, {})", entity, principal, permissions);
       try {
-        internalAccessEnforcer.enforce(entity, principal, permissions);
+        enforcementResult = internalAccessEnforcer.enforce(entity, principal, permissions);
         metricsContext.increment(Constants.Metrics.Authorization.INTERNAL_CHECK_SUCCESS_COUNT, 1);
```

Comment:
It's not nesessarily success.  I think we should differenciate between success/unauthorized/not_needed/failure

Reply:
Do you mean we should 2 more metrics.. like  unauthorized and not_needed [here](https://github.com/cdapio/cdap/blob/65b1aea03e009d561b0c31131cc155125bccc5e3/cdap-common/src/main/java/io/cdap/cdap/common/conf/Constants.java#L1251) 

Or based on enforcementResult's authorized we can increment success or failure

Reply:
yep 2 more metrics.